crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl

```diff
@@ -0,0 +1,214 @@
+    // view, and world coordinates.
+    let Rd_ndc = vec3(frag_coord_to_ndc(in.position).xy, 1.0);
+    let Rd_view = normalize(position_ndc_to_view(Rd_ndc));
+    let Ro_world = view.world_position;
+    let Rd_world = normalize(position_ndc_to_world(Rd_ndc) - Ro_world);
+
+    // Use Beer's law [1] [2] to calculate the maximum amount of light that each
+    // directional light could contribute, and modulate that value by the light
+    // tint and fog color. (The actual value will in turn be modulated by the
+    // phase according to the Henyey-Greenstein formula.)
+    //
+    // We use a bit of a hack here. Conceptually, directional lights are
+    // infinitely far away. But, if we modeled exactly that, then directional
+    // lights would never contribute any light to the fog, because an
+    // infinitely-far directional light combined with an infinite amount of fog
+    // would result in complete absorption of the light. So instead we pretend
+    // that the directional light is `max_depth` units away and do the
+    // calculation in those terms. This simplification also allows us to avoid
+    // shooting secondary rays toward the directional light, which
+    // asymptotically improves performance.
+    //
+    // [1]: https://www.scratchapixel.com/lessons/3d-basic-rendering/volume-rendering-for-developers/intro-volume-rendering.html
+    //
+    // [2]: https://en.wikipedia.org/wiki/Beer%E2%80%93Lambert_law
+    let light_attenuation = exp(-density * max_depth * (absorption + scattering));
+    let light_factors_per_step = fog_color * light_tint * light_attenuation * scattering *
+        density * step_size * light_intensity * exposure;
+
+    // Use Beer's law again to accumulate the ambient light all along the path.
+    var accumulated_color = exp(-end_depth * (absorption + scattering)) * ambient_color *
+        ambient_intensity;
+
+    // Pre-calculate absorption (amount of light absorbed by the fog) and
+    // out-scattering (amount of light the fog scattered away). This is the same
+    // amount for every step.
+    let sample_attenuation = exp(-step_size * density * (absorption + scattering));
+
+    // This is the amount of the background that shows through. We're actually
+    // going to recompute this over and over again for each directional light,
+    // coming up with the same values each time.
+    var background_alpha = 1.0;
+
+    for (var light_index = 0u; light_index < directional_light_count; light_index += 1u) {
+        // Volumetric lights are all sorted first, so the first time we come to
+        // a non-volumetric light, we know we've seen them all.
+        let light = &lights.directional_lights[light_index];
+        if (((*light).flags & DIRECTIONAL_LIGHT_FLAGS_VOLUMETRIC_BIT) == 0) {
+            break;
+        }
+
+        // Offset the depth value by the bias.
+        let depth_offset = (*light).shadow_depth_bias * (*light).direction_to_light.xyz;
```

Comment:
The depth bias is to avoid shadow acne. Google pulls this up: https://digitalrune.github.io/DigitalRune-Documentation/html/3f4d959e-9c98-4a97-8d85-7a73c26145d7.htm